### PR TITLE
Port use_item_on_mon from pokediamond

### PR DESCRIPTION
**Pull request by Admiral-Fish**

### CODE REVIEW

File: src/use_item_on_mon.c [hunk 1]
```diff
@@ -0,0 +1,695 @@
+#include "use_item_on_mon.h"
+#include "item.h"
+#include "pokemon.h"
+#include "party.h"
+#include "constants/moves.h"
+#include "constants/items.h"
+#include "move.h"
+
+BOOL CanUseItemOnPokemon(POKEMON * pokemon, u16 itemID, s32 moveIdx, HeapID heapID)
+{
```
Comment (PikalaxALT) [inline src/use_item_on_mon.c:10]:
```suggestion
BOOL CanUseItemOnPokemon(POKEMON * pokemon, u16 itemID, s32 moveIdx, HeapID heapID) {
```

File: src/use_item_on_mon.c [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+
+    if (GetItemAttr_PreloadedItemData(itemData, ITEMATTR_DEF_EV_UP)) {
+        int dDefEv = GetItemAttr_PreloadedItemData(itemData, ITEMATTR_DEF_EV_UP_PARAM);
+        if (dDefEv > 0) {
+            if (defEv < MAX_EV && (hpEv + atkEv + defEv + speedEv + spAtkEv + spDefEv) < MAX_EV_SUM) {
+                FreeToHeap(itemData);
+                return TRUE;
+            }
+        }
+        else if (dDefEv < 0) {
```
Comment (PikalaxALT) [inline src/use_item_on_mon.c:152]:
```suggestion
        } else if (dDefEv < 0) {
```

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

File: src/use_item_on_mon.c [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+        hp = maxHp;
+    } else {
+        hp = hp + restoration;
+    }
+
+    SetMonData(pokemon, MON_DATA_HP, &hp);
+}
+
+s32 TryModEV(s32 ev, s32 evSum, s32 by)
+{
```
Comment (tgsm) [inline src/use_item_on_mon.c:565]:
```suggestion
s32 TryModEV(s32 ev, s32 evSum, s32 by) {
```

File: src/use_item_on_mon.c [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+    u8 newMaxPp = WazaGetMaxPp(move, ppUp);
+    pp = pp + newMaxPp - maxPp;
+    SetMonData(pokemon, ppUpAttr, &ppUp);
+    SetMonData(pokemon, ppAttr, &pp);
+    
+    return TRUE;
+}
+
+void RestoreMonHPBy(POKEMON *pokemon, u32 hp, u32 maxHp, u32 restoration)
+{
```
Comment (tgsm) [inline src/use_item_on_mon.c:544]:
```suggestion
void RestoreMonHPBy(POKEMON *pokemon, u32 hp, u32 maxHp, u32 restoration) {
```

File: src/use_item_on_mon.c [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+    if (mod < 0) {
+        mod = 0;
+    }
+
+    SetMonData(pokemon, MON_DATA_FRIENDSHIP, &mod);
+    return TRUE;
+}
+
+void HealParty(PARTY * party)
+{
```
Comment (tgsm) [inline src/use_item_on_mon.c:650]:
```suggestion
void HealParty(PARTY * party) {
```

File: src/use_item_on_mon.c [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+    }
+
+    u32 hp = GetMonData(pokemon, MON_DATA_HP, NULL);
+    if ((GetItemAttr_PreloadedItemData(itemData, ITEMATTR_REVIVE) || GetItemAttr_PreloadedItemData(itemData, ITEMATTR_REVIVE_ALL)) && !GetItemAttr_PreloadedItemData(itemData, ITEMATTR_LEVEL_UP)) {
+        if (hp == 0) {
+            FreeToHeap(itemData);
+            return TRUE;
+        }
+    }
+    else if (GetItemAttr_PreloadedItemData(itemData, ITEMATTR_HP_RESTORE) && hp != 0 && hp < GetMonData(pokemon, MON_DATA_MAXHP, NULL)) {
```
Comment (tgsm) [inline src/use_item_on_mon.c:61]:
```suggestion
    } else if (GetItemAttr_PreloadedItemData(itemData, ITEMATTR_HP_RESTORE) && hp != 0 && hp < GetMonData(pokemon, MON_DATA_MAXHP, NULL)) {
```

**Review by tgsm — CHANGES_REQUESTED** [review]

File: src/use_item_on_mon.c [hunk 7]
```diff
@@ -0,0 +1,666 @@
...
+    u8 pp = GetMonData(pokemon, ppAttr, NULL);
+    int ppUpAttr = MON_DATA_MOVE1PPUP + moveIdx;
+    u8 maxPp = WazaGetMaxPp(move_id, GetMonData(pokemon, ppUpAttr, NULL));
+    if (pp < maxPp) {
+        if (ppRestore == PP_RESTORE_ALL) {
+            pp = maxPp;
+        } else {
+            pp += ppRestore;
+            if (pp > maxPp)
+                pp = maxPp;
```
Comment (tgsm) [inline src/use_item_on_mon.c:502]:
```suggestion
            if (pp > maxPp) {
                pp = maxPp;
            }
```

**Review by tgsm — COMMENTED** [review]
Missed this one, sorry. Otherwise looks fine to me

**Review by tgsm — APPROVED** [review]

**Review by PikalaxALT — APPROVED** [review]

**Pull request merged**